File: swift/matrix_multiply.swift
import Foundation

func matrixMultiply(_ a: [[Int]], _ b: [[Int]], _ c: inout [[Int]], _ size: Int) {
    for i in 0..<size {
        for j in 0..<size {
            c[i][j] = 0
            for k in 0..<size {
                c[i][j] += a[i][k] * b[k][j]
            }
        }
    }
}

func allocateMatrix(_ size: Int) -> [[Int]] {
    return Array(repeating: Array(repeating: 0, count: size), count: size)
}

func initMatrixSeeded(_ matrix: inout [[Int]], _ size: Int, _ seedOffset: Int) {
    for i in 0..<size {
        for j in 0..<size {
            matrix[i][j] = (i * 3 + j * 7 + seedOffset) % 100
        }
    }
}

let args = CommandLine.arguments
if args.count < 2 {
    print("Usage: matrix_multiply <size>")
    exit(1)
}

guard let size = Int(args[1]) else {
    print("Invalid number: \(args[1])")
    exit(1)
}

var a = allocateMatrix(size)
var b = allocateMatrix(size)
var c = allocateMatrix(size)

initMatrixSeeded(&a, size, 0)
initMatrixSeeded(&b, size, 1)

let start = CFAbsoluteTimeGetCurrent()
matrixMultiply(a, b, &c, size)
let end = CFAbsoluteTimeGetCurrent()

// Calculate checksum to verify correctness
var checksum: Int64 = 0
for i in 0..<size {
    for j in 0..<size {
        checksum += Int64(c[i][j])
    }
}

let timeMs = Int((end - start) * 1000)

print("Swift: matrix_multiply(\(size)) = \(checksum)")
print("Time: \(timeMs)ms")

File: c/matrix_multiply.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void matrix_multiply(int** a, int** b, int** c, int size) {
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            c[i][j] = 0;
            for (int k = 0; k < size; k++) {
                c[i][j] += a[i][k] * b[k][j];
            }
        }
    }
}

int** allocate_matrix(int size) {
    int** matrix = malloc(size * sizeof(int*));
    for (int i = 0; i < size; i++) {
        matrix[i] = malloc(size * sizeof(int));
    }
    return matrix;
}

void free_matrix(int** matrix, int size) {
    for (int i = 0; i < size; i++) {
        free(matrix[i]);
    }
    free(matrix);
}

void init_matrix(int** matrix, int size, int seed_offset) {
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            matrix[i][j] = (i * 3 + j * 7 + seed_offset) % 100;
        }
    }
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        printf("Usage: matrix_multiply <size>\n");
        return 1;
    }
    
    int size = atoi(argv[1]);
    
    int** a = allocate_matrix(size);
    int** b = allocate_matrix(size);
    int** c = allocate_matrix(size);
    
    init_matrix(a, size, 0);
    init_matrix(b, size, 1);
    
    clock_t start = clock();
    matrix_multiply(a, b, c, size);
    clock_t end = clock();
    
    double time_ms = ((double)(end - start) / CLOCKS_PER_SEC) * 1000;
    
    // Calculate checksum to verify correctness
    long checksum = 0;
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            checksum += c[i][j];
        }
    }
    
    printf("C: matrix_multiply(%d) = %ld\n", size, checksum);
    printf("Time: %.0fms\n", time_ms);
    
    free_matrix(a, size);
    free_matrix(b, size);
    free_matrix(c, size);
    
    return 0;
}